### PR TITLE
Adding extension points for security context.

### PR DESCRIPTION
**Pull request by jboeuf**

### CODE REVIEW

File: src/core/lib/security/context/security_context.h [hunk 1]
```diff
@@ -84,13 +84,19 @@ void grpc_auth_context_unref(grpc_auth_context *policy);
 
 void grpc_auth_property_reset(grpc_auth_property *property);
 
+typedef struct {
```
Comment (markdroth) [inline src/core/lib/security/context/security_context.h]: Please add a comment documenting what this is for and how it will be used.

File: include/grpc++/impl/codegen/client_context.h [hunk 2]
```diff
@@ -307,6 +307,12 @@ class ClientContext {
   };
   static void SetGlobalCallbacks(GlobalCallbacks* callbacks);
 
+  // Should be used for framework-level extensions only.
+  // Applications never need to call this method.
+  grpc_call* c_call() const {
```
Comment (markdroth) [inline include/grpc++/impl/codegen/client_context.h]: Out of curiosity, why is this needed?

This seems reasonable to me, but I'd also like @ctiller to comment, since this is part of the public API.

Comment (jboeuf) [inline include/grpc++/impl/codegen/client_context.h]: In C we can easily extend the call API by exposing such a method:
```
grpc_my_extension* grpc_call_get_my_extension(grpc_call* call);
```

In C++, the `Server/ClientContext` is a class which is much harder to extend. @ctiller and I discussed this already. We *could* have added a macro in the class definition but that felt really yucky. This alternative solution will allow us to do something like:
```
grpc::MyExtension* GetMyExtensionFromServerContext(grpc::ServerContext* ctx);
```
which seems better overall.

**Comment by markdroth** [comment]
LGTM.

Please do add the comment I requested in security_context.h before merging.  Thanks!

**Comment by jboeuf** [comment]
@markdroth Will add the comment for sure thanks!
Also, I think that I should remove the `const` qualifier on the `c_call` method.

**Comment by jboeuf** [comment]
Thanks much for the feedback. PTAL.

**Comment by markdroth** [comment]
LGTM

**Comment by jboeuf** [comment]
Anything holding the merge here? Thanks much!

**Comment by markdroth** [comment]
Let me try re-running the interop tests.  I'm pretty sure that failure was a flake.
